### PR TITLE
Networking

### PR DESCRIPTION
**Pull request by Joerseven**
A few little changes

### CODE REVIEW

File: CSC8503CoreClasses/GameObject.cpp [hunk 1]
```diff
@@ -10,10 +10,11 @@ GameObject::GameObject(string objectName)	{
 	name			= objectName;
 	worldID			= -1;
 	isActive		= true;
-	boundingVolume	= nullptr;
-	physicsObject	= nullptr;
-	renderObject	= nullptr;
-	networkObject	= nullptr;
+	boundingVolume	= nullptr; // Replicated
+    networkObject	= nullptr; // Replicated
```
Comment (Odycious) [inline CSC8503CoreClasses/GameObject.cpp:14]: Tab stuffs, change to spaces :)

File: CSC8503CoreClasses/GameWorld.h [hunk 2]
```diff
@@ -59,16 +60,24 @@ namespace NCL {
...
+            std::vector<NetworkObject*> networkObjects;
+
 			Camera* mainCamera;
 
 			bool shuffleConstraints;
 			bool shuffleObjects;
 			int		worldIDCounter;
 			int		worldStateCounter;
+
+            int networkIdCounter;
```
Comment (Odycious) [inline CSC8503CoreClasses/GameWorld.h:80]: spaces stuff again I guess...

File: CSC8503CoreClasses/NetworkBase.h [hunk 3]
```diff
@@ -11,10 +12,11 @@ enum BasicNetworkMessages {
 	String_Message,
 	Delta_State,	//1 byte per channel since the last state
 	Full_State,		//Full transform etc
-	Received_State, //received from a client, informs that its received packet n
+	Received_State, // Player Input
```
Comment (Odycious) [inline CSC8503CoreClasses/NetworkBase.h:15]: Rename to player_Input

File: CSC8503CoreClasses/NetworkBase.h [hunk 4]
```diff
@@ -11,10 +12,11 @@ enum BasicNetworkMessages {
...
 	Player_Connected,
 	Player_Disconnected,
-	Shutdown
+	Shutdown,
+    Function
```
Comment (Odycious) [inline CSC8503CoreClasses/NetworkBase.h:19]: Spaces

File: CSC8503CoreClasses/State.h [hunk 5]
```diff
@@ -16,6 +16,7 @@ namespace NCL {
 			}
 			
 			virtual void Update(float dt) = 0;
+            virtual void ReceivePacket(int type, GamePacket *payload, int source) { }
```
Comment (Odycious) [inline CSC8503CoreClasses/State.h:19]: spaces

File: CSC8503CoreClasses/StateMachine.cpp [hunk 6]
```diff
@@ -21,27 +21,39 @@ void StateMachine::AddState(State* s) {
 	allStates.emplace_back(s);
 	if (activeState == nullptr) {
 		activeState = s;
+        activeState->OnEnter();
```
Comment (Odycious) [inline CSC8503CoreClasses/StateMachine.cpp:24]: spaces

File: Client/Client.cpp [hunk 7]
```diff
@@ -2,5 +2,57 @@
...
+    stateManager = std::make_unique<StateMachine>();
+    world = std::make_unique<GameWorld>();
+    renderer = std::make_unique<GameTechRenderer>(*world);
+
+    InitStateManager();
+    RegisterPackets();
+}
+
+void Client::InitStateManager() {
+    auto clientMenu = new MenuState(renderer.get(), world.get(), baseClient.get());
```
Comment (Odycious) [inline Client/Client.cpp:21]: Possibly rename baseClient?

File: Client/States/GameplayState.cpp [hunk 8]
```diff
@@ -0,0 +1,123 @@
+#include "GameplayState.h"
+
+using namespace NCL;
+using namespace CSC8503;
+
+GameplayState::GameplayState(GameTechRenderer* pRenderer, GameWorld* pGameworld, GameClient* pClient) : State() {
+	renderer = pRenderer;
+	world = pGameworld;
+    baseClient = pClient;
```
Comment (Odycious) [inline Client/States/GameplayState.cpp:9]: spaces

File: Client/States/GameplayState.cpp [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+GameplayState::~GameplayState() {}
+
+void GameplayState::OnEnter() {
+
+    Window::GetWindow()->ShowOSPointer(false);
+    Window::GetWindow()->LockMouseToWindow(true);
+
+    firstPersonPosition = nullptr;
+	InitialiseAssets();
```
Comment (Odycious) [inline Client/States/GameplayState.cpp]: spaces

File: Client/States/GameplayState.cpp [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+void GameplayState::Update(float dt) {
+
+    if (firstPersonPosition) {
+        world->GetMainCamera()->SetPosition(firstPersonPosition->GetPosition());
+    }
+
+	world->GetMainCamera()->UpdateCamera(dt);
+    SendInputData();
```
Comment (Odycious) [inline Client/States/GameplayState.cpp:35]: space platz

File: Client/States/GameplayState.cpp [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    input.playerDirection = playerDirection;
+
+    baseClient->SendPacket(input);
+}
+
+
+void GameplayState::InitialiseAssets() {
+	InitCamera();
+	InitWorld();
+    FinishLoading();
```
Comment (Odycious) [inline Client/States/GameplayState.cpp:75]: platz

File: Client/States/GameplayState.cpp [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+
+void GameplayState::InitWorld() {
+    CreatePlayers();
+}
+
+void GameplayState::CreatePlayers() {
+    for (int i=0; i<Replicated::PLAYERCOUNT; i++) {
+        auto player = new GameObject();
+        replicated->CreatePlayer(player, *world);
+        player->SetRenderObject(new RenderObject(&player->GetTransform(), resources->GetMesh("Goat.msh"), nullptr, nullptr));
```
Comment (Odycious) [inline Client/States/GameplayState.cpp:98]: #Swap that silly goat out for an EVIL GENIUS!!

File: Client/States/GameplayState.cpp [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+    }
+}
+
+
+bool GameplayState::IsDisconnected() {
```
Comment (Odycious) [inline Client/States/GameplayState.cpp:103]: In progress

File: Client/States/MenuState.cpp [hunk 14]
```diff
@@ -0,0 +1,75 @@
+#include "MenuState.h"
+using namespace NCL;
+using namespace CSC8503;
+
+MenuState::MenuState(GameTechRenderer* pRenderer, GameWorld* pGameworld, GameClient* pClient) : State()
+{
+    renderer = pRenderer;
+	world = pGameworld;
```
Comment (Odycious) [inline Client/States/MenuState.cpp]: spaces

**Review by DelCherub — APPROVED** [review]
Lots of stuff here but it looks alright to me

File: Replicated/PacketTypes.h [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    int functionId;
+    FunctionData data{};
+    FunctionPacket(int id, FunctionData* d) {
+        if (d) {
+            // Not sure if copy constructor works here, test afterward.
+            memcpy(&data, d, sizeof(FunctionData));
+        }
+        type = Function;
+        functionId = id;
+        size = sizeof(FunctionPacket) - sizeof(GamePacket) + 20;
```
Comment (Odycious) [inline Replicated/PacketTypes.h]: FLAT 20 -- Bad bad not good

File: Replicated/Replicated.cpp [hunk 16]
```diff
@@ -3,3 +3,25 @@
 //
 
 #include "Replicated.h"
+
+
+
+Replicated::Replicated() {
+}
+
+void Replicated::CreateObject(GameObject *g, GameWorld& world) {
```
Comment (Odycious) [inline Replicated/Replicated.cpp:12]: In progress

File: Replicated/Replicated.cpp [hunk 17]
```diff
@@ -3,3 +3,25 @@
...
+
+void Replicated::CreatePlayer(GameObject *g, GameWorld& world) {
+    constexpr float meshSize = 1.0f;
+    world.AddGameObject(g, true);
+    auto volume = new AABBVolume(Vector3(meshSize, meshSize, meshSize));
+    g->SetBoundingVolume((CollisionVolume*)volume);
+
+    g->GetTransform()
+            .SetScale(Vector3(meshSize, meshSize, meshSize))
+            .SetPosition(Vector3(0 + (g->GetWorldID()%2) * 10,0,10 * (g->GetWorldID()/2)));
```
Comment (Odycious) [inline Replicated/Replicated.cpp:24]: For testing purposes, to see different objects distinctly.

File: Server/GameServer.cpp [hunk 18]
```diff
@@ -0,0 +1,121 @@
+#include "GameServer.h"
+#include "enet.h"
+
+using namespace NCL;
+using namespace CSC8503;
+
+GameServer::GameServer(int onPort, int maxClients)	{
```
Comment (Odycious) [inline Server/GameServer.cpp:7]: tab after () change to spaces

File: Server/GameServer.cpp [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+    Shutdown();
+}
+
+void GameServer::CallRemote(int functionId, FunctionData* data, int peer) {
+    // Can later optimize this to adjust packet size based on argument size.
+    FunctionPacket packet(functionId, data);
+    SendPacket(packet, peer);
+}
+
+void GameServer::CallRemoteAll(int functionId, FunctionData* data) {
```
Comment (Odycious) [inline Server/GameServer.cpp:27]: Maybe add some comments to remind people how to use the funcs. E.g Call remote from server to client or client to server or both?

File: Server/Server.h [hunk 20]
```diff
@@ -5,8 +5,58 @@
...
+    std::unique_ptr<GameServer> serverBase;
+    std::unique_ptr<GameWorld> world;
+    std::unique_ptr<PhysicsSystem> physics;
+    std::unique_ptr<Replicated> replicated;
+
+    float packetTimer;
+    int sceneSnapshotId;
+
+    void RegisterPackets();
+//    void InitGame();
```
Comment (Odycious) [inline Server/Server.h]: Delete commented out stuff

**Review by Odycious — COMMENTED** [review]
That was a lot to read, my brain hurts but good work! Networking diff.

**Pull request merged**